pkg/restapi/device/operations.go

```diff
@@ -0,0 +1,214 @@
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package device
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+
+	"github.com/trustbloc/edge-agent/pkg/restapi/common/store"
+	"github.com/trustbloc/edge-agent/pkg/restapi/common/store/cookie"
+	"github.com/trustbloc/edge-agent/pkg/restapi/common/store/tokens"
+	"github.com/trustbloc/edge-agent/pkg/restapi/common/store/user"
+
+	"github.com/duo-labs/webauthn.io/session"
+	"github.com/duo-labs/webauthn/protocol"
+	"github.com/duo-labs/webauthn/webauthn"
+	"github.com/trustbloc/edge-core/pkg/log"
+	"github.com/trustbloc/edge-core/pkg/storage"
+
+	"github.com/trustbloc/edge-agent/pkg/restapi/common"
+)
+
+// Endpoints.
+const (
+	registerBeginPath  = "/register/begin/"
+	registerFinishPath = "/register/finish"
+)
+
+// Stores.
+const (
+	transientStoreName = "edgeagent_device_rgx"
+	deviceCookieName   = "device_user"
+)
+
+var logger = log.New("edge-agent/device-registration")
+
+// Config holds all configuration for an Operation.
+type Config struct {
+	Storage    *StorageConfig
+	UIEndpoint string
+	TLSConfig  *tls.Config
+	Keys       *KeyConfig
+}
+
+type KeyConfig struct {
+	Auth []byte
+	Enc  []byte
+}
+
+// StorageConfig holds storage config.
+type StorageConfig struct {
+	Storage          storage.Provider
+	TransientStorage storage.Provider
+}
+
+type stores struct {
+	users     *user.Store
+	tokens    *tokens.Store
+	transient storage.Store
+	cookies   cookie.Store
+}
+
+// Operation implements OIDC operations.
+type Operation struct {
+	store        *stores
+	uiEndpoint   string
+	tlsConfig    *tls.Config
+	sessionStore *session.Store
+	webauthn    *webauthn.WebAuthn
+}
+
+// New returns a new Operation.
+func New(config *Config) (*Operation, error) {
+	op := &Operation{
+		store: &stores{
+			cookies: cookie.NewStore(config.Keys.Auth, config.Keys.Enc),
+		},
+		uiEndpoint: config.UIEndpoint,
+		tlsConfig:  config.TLSConfig,
+	}
+
+	var err error
+
+	op.store.transient, err = store.Open(config.Storage.TransientStorage, transientStoreName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open transient store: %w", err)
+	}
+
+	op.sessionStore, err = session.NewStore()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create session store: %w", err)
+	}
+
+	op.store.tokens, err = tokens.NewStore(config.Storage.Storage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open tokens store: %w", err)
+	}
+
+	return op, nil
+}
+
+// GetRESTHandlers get all controller API handler available for this service.
+func (o *Operation) GetRESTHandlers() []common.Handler {
+	return []common.Handler{
+		common.NewHTTPHandler(registerBeginPath, http.MethodGet, o.beginRegistration),
+		common.NewHTTPHandler(registerFinishPath, http.MethodGet, o.finishRegistration),
+	}
+}
+
+func (o *Operation) beginRegistration(w http.ResponseWriter, r *http.Request) {
+	logger.Debugf("handling device registration: %s", r.URL.String())
+
+	userSub := r.URL.Query().Get("user")
```

Comment:
We should be fetching the user's identifier from session cookies. The session cookie should be set up at the end of `oidc.Operation.oidcCallbackHandler()`. The user's identifier will not form part of the request URL.

Opened #419 for this.

Reply:
Is it okay if i address this in a separate follow up issue ?

Reply:
done